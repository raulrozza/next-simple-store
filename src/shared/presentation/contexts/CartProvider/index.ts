export { default as useAddToCart } from './hooks/useAddToCart';
