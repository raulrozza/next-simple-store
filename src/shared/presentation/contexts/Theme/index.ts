export { default as useTheme } from './hooks/useTheme';
export { default as Theme } from './implementations/StyledComponentTheme';
