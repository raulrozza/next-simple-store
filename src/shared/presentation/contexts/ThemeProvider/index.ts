export { default as useTheme } from './hooks/useTheme';
export { default as ThemeProvider } from './implementations/StyledComponentTheme';
