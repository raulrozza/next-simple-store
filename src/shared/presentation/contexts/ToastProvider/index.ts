export { useToastProvider } from './context';
export { default as ToastProvider } from './implementations/ReactToastifyProvider';
