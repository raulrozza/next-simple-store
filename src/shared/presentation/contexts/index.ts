export * from './CartProvider';
export * from './ThemeProvider';
