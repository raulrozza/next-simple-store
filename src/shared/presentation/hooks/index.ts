import { trpc } from '@/shared/infra/http/trpc';

const { useQuery, useMutation } = trpc;

export { useQuery, useMutation };
