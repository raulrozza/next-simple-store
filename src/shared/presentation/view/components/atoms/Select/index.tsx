import { FC } from 'react';

import { useField } from 'formik';

import { ISelect } from '@/shared/presentation/entities/Field';

import { ErrorContainer, StyledSelect } from './styles';

const Select: FC<Omit<ISelect, 'type'>> = ({
    name,
    options,
    placeholder,
    ...props
}) => {
    const [field, meta] = useField(name);

    const hasErrors = Boolean(meta.touched && meta.error);

    return (
        <div>
            <StyledSelect {...props} {...field} hasErrors={hasErrors}>
                {placeholder && <option value="">{placeholder}</option>}

                {options.map(({ value, label }) => (
                    <option key={value} value={value}>
                        {label}
                    </option>
                ))}
            </StyledSelect>

            {hasErrors && <ErrorContainer>{meta.error}</ErrorContainer>}
        </div>
    );
};

export default Select;
