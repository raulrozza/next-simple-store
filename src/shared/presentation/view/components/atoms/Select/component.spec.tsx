import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Formik } from 'formik';

import { ThemeProvider } from '@/shared/presentation/contexts';

import Select from './index';

describe('Select', () => {
    it('should render the select correctly with all the options', () => {
        const view = render(
            <Formik
                initialValues={{
                    select: '',
                }}
                onSubmit={jest.fn()}
            >
                <Select
                    name="select"
                    placeholder="My select"
                    options={[
                        { value: 'opt1', label: 'option 1' },
                        { value: 'opt2', label: 'option 2' },
                    ]}
                    data-testid="select-input"
                />
            </Formik>,
            {
                wrapper: ThemeProvider,
            },
        );

        expect(screen.getByTestId('select-input')).toBeInTheDocument();
        expect(screen.getByText('My select')).toBeInTheDocument();
        expect(screen.getByText('option 1')).toBeInTheDocument();
        expect(screen.getByText('option 2')).toBeInTheDocument();
        expect(view).toMatchSnapshot();
    });

    describe('Value handling', () => {
        it('should init the select with the initial field value', () => {
            const view = render(
                <Formik initialValues={{ select: 'opt2' }} onSubmit={jest.fn()}>
                    <Select
                        name="select"
                        options={[
                            { value: 'opt1', label: 'option 1' },
                            { value: 'opt2', label: 'option 2' },
                        ]}
                        data-testid="select-input"
                    />
                </Formik>,
                { wrapper: ThemeProvider },
            );

            expect(screen.getByTestId('select-input')).toHaveValue('opt2');
            expect(view).toMatchSnapshot();
        });

        it('should change the field value when changing the selection', async () => {
            const user = userEvent.setup();

            render(
                <Formik initialValues={{ select: 'opt1' }} onSubmit={jest.fn()}>
                    <Select
                        name="select"
                        options={[
                            { value: 'opt1', label: 'option 1' },
                            { value: 'opt2', label: 'option 2' },
                        ]}
                        data-testid="select-input"
                    />
                </Formik>,
                { wrapper: ThemeProvider },
            );

            const select = screen.getByTestId('select-input');
            expect(select).toHaveValue('opt1');

            await user.selectOptions(select, 'opt2');

            expect(select).toHaveValue('opt2');
        });
    });

    describe('Error handling', () => {
        it('should show the error message if the select has errors and has been touched', async () => {
            const user = userEvent.setup();

            render(
                <Formik
                    initialValues={{
                        select: '',
                    }}
                    validate={() => ({ select: 'My error' })}
                    onSubmit={jest.fn()}
                    validateOnChange
                >
                    <Select
                        name="select"
                        placeholder="My select"
                        options={[
                            {
                                value: 'new value',
                                label: 'great new value',
                            },
                        ]}
                        data-testid="select-input"
                    />
                </Formik>,
                {
                    wrapper: ThemeProvider,
                },
            );

            expect(screen.queryByText('My error')).not.toBeInTheDocument();

            await user.selectOptions(
                screen.getByTestId('select-input'),
                'new value',
            );
            act(() => {
                screen.getByTestId('select-input').blur(); // Deselect to trigger touch
            });

            expect(await screen.findByText('My error')).toBeInTheDocument();
        });

        it('should not display the error message if the select was not touched', async () => {
            const user = userEvent.setup();

            render(
                <Formik
                    initialValues={{ select: 'my value' }}
                    validate={() => ({ select: 'my error' })}
                    onSubmit={jest.fn()}
                    validateOnChange
                >
                    <Select
                        name="select"
                        placeholder="My select"
                        options={[
                            {
                                value: 'new value',
                                label: 'great new value',
                            },
                        ]}
                        data-testid="select-input"
                    />
                </Formik>,
                { wrapper: ThemeProvider },
            );

            expect(screen.queryByText('My error')).not.toBeInTheDocument();

            await user.selectOptions(
                screen.getByTestId('select-input'),
                'new value',
            );

            expect(screen.queryByText(/my error/)).toBeNull();
        });
    });
});
