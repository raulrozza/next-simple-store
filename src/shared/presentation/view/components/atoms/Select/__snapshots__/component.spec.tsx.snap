// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Value handling should init the select with the initial field value 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  background-color: #FFFFFF;
  border: 2px solid #E6EAEA;
  outline: none;
  border-radius: 0.25rem;
  width: 100%;
  padding: 0.5rem;
  -webkit-transition: outline 0.2s,border 0.2s;
  transition: outline 0.2s,border 0.2s;
}

.c0:focus {
  border-color: #8CCFDE;
}

<body>
    <div>
      <div>
        <select
          class="c0"
          data-testid="select-input"
          name="select"
        >
          <option
            value="opt1"
          >
            option 1
          </option>
          <option
            value="opt2"
          >
            option 2
          </option>
        </select>
      </div>
    </div>
  </body>,
  "container": .c0 {
  background-color: #FFFFFF;
  border: 2px solid #E6EAEA;
  outline: none;
  border-radius: 0.25rem;
  width: 100%;
  padding: 0.5rem;
  -webkit-transition: outline 0.2s,border 0.2s;
  transition: outline 0.2s,border 0.2s;
}

.c0:focus {
  border-color: #8CCFDE;
}

<div>
    <div>
      <select
        class="c0"
        data-testid="select-input"
        name="select"
      >
        <option
          value="opt1"
        >
          option 1
        </option>
        <option
          value="opt2"
        >
          option 2
        </option>
      </select>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Select should render the select correctly with all the options 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  background-color: #FFFFFF;
  border: 2px solid #E6EAEA;
  outline: none;
  border-radius: 0.25rem;
  width: 100%;
  padding: 0.5rem;
  -webkit-transition: outline 0.2s,border 0.2s;
  transition: outline 0.2s,border 0.2s;
}

.c0:focus {
  border-color: #8CCFDE;
}

<body>
    <div>
      <div>
        <select
          class="c0"
          data-testid="select-input"
          name="select"
        >
          <option
            value=""
          >
            My select
          </option>
          <option
            value="opt1"
          >
            option 1
          </option>
          <option
            value="opt2"
          >
            option 2
          </option>
        </select>
      </div>
    </div>
  </body>,
  "container": .c0 {
  background-color: #FFFFFF;
  border: 2px solid #E6EAEA;
  outline: none;
  border-radius: 0.25rem;
  width: 100%;
  padding: 0.5rem;
  -webkit-transition: outline 0.2s,border 0.2s;
  transition: outline 0.2s,border 0.2s;
}

.c0:focus {
  border-color: #8CCFDE;
}

<div>
    <div>
      <select
        class="c0"
        data-testid="select-input"
        name="select"
      >
        <option
          value=""
        >
          My select
        </option>
        <option
          value="opt1"
        >
          option 1
        </option>
        <option
          value="opt2"
        >
          option 2
        </option>
      </select>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
